Specify account linking

tung2744 opened the pull request:
ref #4066

tung2744 commented:
@louischan-oursky Updated according to our latest discussion, thanks!

louischan-oursky commented on docs/specs/idp-account-linking.md:
```diff
@@ -12,47 +12,60 @@ Here is an example of the configuration:
+        client_id: exampleclientid
+        type: google
+        account_linking:
+          enabled: false
```
What does `enabled: false` mean? Remember we have to be back-forward compatible. That is, the default config should match the current behavior, that is, error in case of conflict.

If `enabled: false` means `account_linking.oauth=error`, it could be confusing.

tung2744 replied:
@louischan-oursky `enabled: false` means the oauth account will never be considered as matching an existing identity.

So I guess the default should be `enabled: true` with `email` for some of the provider, but some of them can still remain `enabled: false` as default such as wechat. Maybe let me list out the appropriate default values.

tung2744 replied:
```
link_by_claim: "email" # nullable
```

louischan-oursky commented on docs/specs/idp-account-linking.md:
```diff
@@ -12,47 +12,60 @@ Here is an example of the configuration:
 
 - Whenever an oauth identity is used, and there is no existing user for that identity.
 
   1. If `account_linking.enabled` is not `true`, do nothing. Else,
```
Step 1 is not backward compatible. It is because account_link.enabled is false by default. The backward compatible behavior should be `"error"`.

tung2744 replied:
For backward compatibility, modify the default value as mentioned in https://github.com/authgear/authgear-server/pull/4096#discussion_r1555413513

louischan-oursky commented on docs/specs/idp-account-linking.md:
```diff
@@ -12,47 +12,60 @@ Here is an example of the configuration:
-        match_against_claim_key: "email"
+  on_conflict:
+    account_linking:
+      oauth: "merge"
```
I guess the key `oauth` means From OAuth. That is, this config is about the scenario that the triggering identity is an OAuth Identity.

This config, I believe, can handle the following cases

1. Triggering=OAuth, existing=OAuth
2. Triggering=OAuth, existing=Login ID

Then, I wonder why do we need to separate the config of From OAuth and From Login ID, because if we did not make the distinction, the cases would be

1. Triggering=OAuth, existing=OAuth
2. Triggering=OAuth, existing=Login ID
3. Triggering=Login ID, existing=OAuth (The handling of this case should be the same as case 1)

tung2744 replied:
> Triggering=Login ID, existing=OAuth (The handling of this case should be the same as case 1)

I doubt if it is the case. We probably want to block and ask the user to login with oauth if this is this case, while we want to automatically merge accounts in case 1 and 2.

tung2744 replied:
Proposed change:

```
identity:
  on_conflict:
    signup:
      oauth: "merge" # "error" | "merge"
      login_id: "error" # "error" is the only supported value at the moment.
```

tung2744 commented:
- Renamed some config keys
- Use pointer to point to a claim
- Changed default `link_by.pointer` to `/email`
- Added more options of `identity.on_conflict.signup` and documented their corresponding behavior

tung2744 commented on docs/specs/idp-account-linking.md:
```diff
@@ -12,47 +12,74 @@ Here is an example of the configuration:
-  - 1.1. Use `account_linking.match_against_claim_key` to obtain a value from the new user. If the value is empty, do nothing.
-  - 1.2. Use `account_linking.idp_claim_key` to query users with an oauth identity of the current provider matching the value in 1.1.
-  2. (TBC) If there is at least one user found in 1, block the login and ask user to signin with with the found identity.
+- Whenever a new user creates an new identity during signup.
```
> How is this section different from the section starting from line 46?

Basically they are same, this spec separated the two case to show that this steps can handle these two cases.

IniZio commented:
Looks like we have accounting_linking design in login flow, need to update as well?

Can find `account_linking:` in https://github.com/authgear/authgear-server/blob/main/docs/specs/authentication-flow.md#loginflow

tung2744 commented:
I think I will remove that from the current authflow spec, because it differs from the current spec.

tung2744 commented:
Some updates was made to the spec:

- Refined the `link_by` section:
    - Separated the config for matching existing authgear accounts and the external idp claims.
        - Use `identity` to match existing authgear users.
        - Use `idp_claim` to match a claim inside the external idp user info.
    - Removed the `pointer` key.
    - Document the default (Or auto) behavior.
- Added the Login and Link Flow section, which describes what exactly the "login_and_link" action will do.
- Added an Q&A section, which written down some questions and answers we've discussed in meetings.

Maybe @chpapa can also take a look if you got time, thanks!

tung2744 commented:
Further updated the provider configs.

louischan-oursky commented on docs/specs/idp-account-linking.md:
````diff
@@ -0,0 +1,528 @@
+                    - type: view_recovery_code
+```
+
+7. As the next step of the original flow is create primary password authenticator, the user will need to created primary authenticator.
````
to create

louischan-oursky commented on docs/specs/idp-account-linking.md:
```diff
@@ -0,0 +1,528 @@
+
+## Q&A
+
+- Why we need to login the user before linking the account?
```
Use ### for each question. And use a paragraph for each answer.

```
### Why ...

Answer in a paragraph.
```

Remember to re-generate the TOC.

louischan-oursky commented on docs/specs/authentication-flow.md:
```diff
@@ -275,34 +275,6 @@ login_flows:
   # If the condition does not hold, this step is no-op.
   - type: change_password
     target_step: step1
-
```
We still have some account linking content in this document. Please remove them. Also add a backlink to account-linking.md to tell the reader to find the new document.

tung2744 commented:
Updated again for some config renaming and also documented account linking of login ids, as discussed in the product meeting.

tung2744 commented:
Added a spec to document the current supported sso providers

louischan-oursky commented on docs/specs/account-linking.md:
```diff
@@ -0,0 +1,421 @@
+        client_id: exampleclientid
+        type: adfs
+        user_profile_mapping:
+          - from_claim:
```
It should be `oauth_claim`

louischan-oursky commented on docs/specs/account-linking.md:
```diff
@@ -0,0 +1,421 @@
+        user_profile_mapping:
+          - from_claim:
+              pointer: "/primary_phone"
+            to_attribute:
```
It should be `user_profile_attribute`.

chpapa replied:
actually `user_profile` might be good enough? As we defined `user_profile` = standard attributes + custom attributes elsewhere already.

louischan-oursky commented on docs/specs/sso-providers.md:
```diff
@@ -0,0 +1,216 @@
+1. Map the following fields from the response into the user profile.
+
+- From `email` to `email`
+- From `first_name` to `first_name`
```
to `given_name`.

louischan-oursky commented on docs/specs/sso-providers.md:
```diff
@@ -0,0 +1,216 @@
+
+- From `email` to `email`
+- From `first_name` to `first_name`
+- From `last_name` to `last_name`
```
to `family_name`.
